rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+/// * `FnOnce(InputType) -> Future<OutputType>`
+/// * `FnOnce(InputType, Extension<Arc<T>>) -> Future<OutputType>`
+///
+/// Wrapping the service with a [`tower::Layer`] will allow to have operations' signatures with and without shared state:
+///
+/// ```compile_fail
+/// use std::sync::Arc;
+/// use aws_smithy_http_server::{AddExtensionLayer, Extension, Router};
+/// use tower::ServiceBuilder;
+/// use tokio::sync::RwLock;
+///
+/// // Shared state,
+/// #[derive(Debug, State)]
+/// pub struct State {
+///     pub count: RwLock<u64>
+/// }
+///
+/// // Operation implementation with shared state.
+/// async fn operation_with_state(input: Input, state: Extension<Arc<State>>) -> Output {
+///     let mut count = state.0.write().await;
+///     *count += 1;
+///     Ok(Output::new())
+/// }
+///
+/// // Operation implementation without shared state.
+/// async fn operation_without_state(input: Input) -> Output {
+///     Ok(Output::new())
+/// }
+///
+/// let app: Router = OperationRegistryBuilder::default()
+///     .operation_with_state(operation_with_state)
+///     .operation_without_state(operation_without_state)
+///     .build()
+///     .unwrap()
+///     .into();
+/// let shared_state = Arc::new(State::default());
+/// let app = app.layer(ServiceBuilder::new().layer(AddExtensionLayer::new(shared_state)));
+/// let server =
+///     axum::Server::bind(&"0.0.0.0:13734".parse().unwrap()).serve(app.into_make_service());
+/// ...
+/// ```
+///
+/// Without the middleware layer, the framework will require operations' signatures without
+/// the shared state.
+///
+/// [`middleware`]: [`aws_smithy_http_server::AddExtensionLayer`]
+#[derive(Debug)]
+pub struct State {
+    pokemons_translations: HashMap<String, PokemonTranslations>,
+    call_count: AtomicU64,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        let mut pokemons_translations = HashMap::new();
+        pokemons_translations.insert(
+            String::from("pikachu"),
+            PokemonTranslations::new(
+                PIKACHU_ENGLISH_FLAVOR_TEXT,
+                PIKACHU_SPANISH_FLAVOR_TEXT,
+                PIKACHU_ITALIAN_FLAVOR_TEXT,
+            ),
+        );
+        Self {
+            pokemons_translations,
+            call_count: Default::default(),
+        }
+    }
+}
+
+/// Retrieves information about a Pokémon species.
+pub async fn get_pokemon_species(
+    input: input::GetPokemonSpeciesInput,
+    state: Extension<Arc<State>>,
+) -> Result<output::GetPokemonSpeciesOutput, error::GetPokemonSpeciesError> {
+    state.0.call_count.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+    // We only support retrieving information about Pikachu.
+    let pokemon = state.0.pokemons_translations.get(&input.name);
+    match pokemon.as_ref() {
+        Some(pokemon) => {
+            debug!("Requested Pokémon is {}", input.name);
+            let english_flavor_text = model::FlavorText::builder()
```

Comment:
Now that we respect `required`, can we refactor this to not make use of builders?